spec/ParseQuery.spec.js

```diff
@@ -3210,6 +3232,47 @@ describe('Parse.Query testing', () => {
       );
   });
 
+  it('select keys', async () => {
+    const obj = new TestObject({ foo: 'baz', hello: 'world' });
+    await obj.save();
+
+    const response = await request({
+      url: Parse.serverURL + '/classes/TestObject',
+      qs: {
+        keys: 'hello',
+        where: JSON.stringify({ objectId: obj.id }),
+      },
+      headers: masterKeyHeaders,
+    });
+    expect(response.data.results[0].foo).toBeUndefined();
+    expect(response.data.results[0].hello).toBe('world');
+
+    const response2 = await request({
+      url: Parse.serverURL + '/classes/TestObject',
+      qs: {
+        keys: ['foo', 'hello'],
+        where: JSON.stringify({ objectId: obj.id }),
+      },
+      headers: masterKeyHeaders,
+    });
+    expect(response2.data.results[0].foo).toBe('baz');
+    expect(response2.data.results[0].hello).toBe('world');
+
+    const response3 = await request({
+      url: Parse.serverURL + '/classes/TestObject',
+      qs: {
+        keys: [''],
```

Comment:
Why does it have to pass an array element with an empty string and not just an empty array? I think `keys:[]` should yield the same as `keys:['']`, because `keys:['']` is essentially the same as `keys:'keyThatDoesntExist']`.

I see `keys:[]` as a valid extreme case and `keys:['']` as an edge case or maybe even semantically invalid, because no DBMS allows a key that is an empty string. I think this could throw at the DB level and therefore be removed as invalid key and never reach the DB. What we want to test with `keys:['']` is therefore that we clean invalid keys gracefully and end up with `keys:[]` that is sent to the DB.

I see below that you added a test for `excludeKeys:[]`, so I think we could add `keys:[]` as well, and conversely maybe `excludeKeys:['']` as well. Sorry, I don't mean to blow up this PR, but since this empty string seems at the center of discussion here, maybe it makes sense to add.

Reply:
> Why does it have to pass an array element with an empty string and not just an empty array? I think keys:[] should yield the same as keys:[''], because keys:[''] is essentially the same as keys:['keyThatDoesntExist'].

I pass `keys:['']` to replicate the current/original functionality on the main branch. I think you are expecting that the original functionality behaves one way, but it does not. I pointed out the JS, Flutter, and Android SDKs are sending a delineated string instead of a JSON string array, so it makes sense what works in REST vs JS is different. You are expecting them to be same. If you copy this test and run it on the original code, it will fail, what will pass is `keys:['']`. Passing `keys:[]` will return all keys of an object.

If you are suggesting I add the functionality of `keys:[]` through REST, then that's a different discussion and PR as we discussed on the thread as: 1) That's not part of the original functionality 2) I don't even agree it should be there in REST or any other SDK (per my comments on the thread)... What I did agree with was adding back the original functionality (even though I don't like it) in order to not break what devs have already been doing through the SDKs using `keys: []`, but also add the ability to digest an encoded JSON string array (what it should have been doing in the first place). If all of the original designs were around a JSON string array, then what you are expecting would probably work, unfortunately this isn't the case.

> I see keys:[] as a valid extreme case and keys:[''] as an edge case or maybe even semantically invalid, because no DBMS allows a key that is an empty string.

I think my comments above answers this as there's a difference between what you expect it to do and what it actually does. Making this work as expected will probably break the SDKs or require hacky code to allow both the original and expected to work together.

> I see below that you added a test for excludeKeys:[], so I think we could add keys:[] as well, and conversely maybe excludeKeys:['']

What are your expectations from these tests? My interpretations of what you have been asking is that you believe `keys` and `excludeKeys` are inverses of each other (with some type of mix of them doing the same thing at times, e.g. select no keys acts as exclude all keys). Though that seems correct in theory (excluding the mix part I mentioned and keep saying I don't agree with), it's not how they are currently coded, there are differences.  

If you want to move towards what you expect, I suggest the following steps:

1. This PR can remain as is, I can add a couple more tests, but I'm sure the tests won't yield what you expect.
2. Someone can create a PR to update the JS SDK to send arrays, update the test cases, accept the PR, bump the JS version, then use that new JS version in Parse Server... This will need to be done in order to test properly and compare against what's actually going on and what you expect. Note that my guess is when you make the changes on the Server side to work with your expectations, your changes will probably break the Flutter and Android SDKs, the Parse-Swift SDK will most likely work fine.
3. Propagate the changes to the update to the Flutter and Android SDKs

Or, my second suggestion is to:

1. In this PR, only make this PR work as what is expected with REST, which will probably break the functionality of the JS, Flutter, and Android SDKs, Parse-Swift will be fine. You will also have to `xit` some of the JS tests on the Parse Server as I don't see how you will get them to work without changing the JS SDK under the hood.

Lastly:
1. This PR can remain as is, I can add a couple more tests, but I'm sure they will be different from what you expect.
2. All of the SDKs can proceed as they were, but the behavior you expect won't be there until someone provides the fix and propagates them to the other SDKs

Reply:
> I see below that you added a test for excludeKeys:[], so I think we could add keys:[] as well, and conversely maybe excludeKeys:[''] as well.

I just added the test cases for these. I believe if you look at the `expect` of the testcases and compare them to what you expect to happen, you will see what I was talking about above. I also ran these same exact test cases on the original branch. All pass except for the one that passes an array of `excludeKeys: ['foo', 'hello']` in REST as the original code didn't support that and returned all keys instead.

Reply:
Hm - I also prefer to keep this PR as small as possible, as you suggested.

> what works in REST vs JS is different

I didn't realize that. I thought that this PR would change that Parse Server accepts both, a delimited string OR an array. And that via the JS SDK methods as well as via REST API. If that was the change, wouldn't that solve the original issue?

Reply:
> I didn't realize that. I thought that this PR would change that Parse Server accepts both, a delimited string OR an array. And that via the JS SDK methods as well as via REST API. Wouldn't that solve the original issue?

The part that's fixed, I haven't been able to figure out how to write a test case for, because it will require the JS SDK to send a JSON string array for proper testing on the Server. If you recall yesterday, I mentioned I would ping back to make sure the changes still work with Parse-Swift client. I did this by running Parse-Swift in Swift Playgrounds against a server docker image that contained the PR changes. I know what the client sends and I know the server digests JSON string arrays properly in many places, but discovered it does it differently for `keys` and `excludeKeys`. The Parse-Swift SDK sends array strings using REST the exact same way for everything, PUT, POST, GET, etc. and has test cases for what it encodes to the server.
 
Note that my original fix:
```javascript
if (body.keys) {
      options.keys = String(body.keys);
}
``` 

made the server work with either, JSON string array or comma joined string, but causes `keys: []` in JS and `keys: ['']` to not work anymore. This lead me to believe that, `keys: []` from the JS SDK is converted to either `keys: ['']` or `keys: ''` (more likely) in REST, which I can believe as I can see `[]` being converted to an empty string, making my original fix above fail. The current changes allow for all of the old functionality along with a working client that sends an JSON string array of `[key1, key2]`.

Once the JS SDK is patched, I think it will allow proper testing and ensure that breaks in the future are detected. My guess is that a fix to the JS SDK, may not be backwards compatible with older servers though...

Reply:
The test cases added in [commit](https://github.com/parse-community/parse-server/pull/7242/commits/0b72a7a8181b3d86aff70076abf3630d421d08d8) properly represent when a client sends a JSON array string using REST (which is what Parse-Swift sends). If you run the whole select REST test case on the original main branch, it will fail in a few places

Reply:
The original code will fail in the following (which all pass in the PR):

```bash
Failures:
1) Parse.Query testing select keys
  Message:
    Expected 1 to be undefined.
  Stack:
    Error: Expected 1 to be undefined.
        at <Jasmine>
        at UserContext.<anonymous> (/Users/coreybaker/Documents/github/parse-server/spec/ParseQuery.spec.js:3260:43)
        at <Jasmine>
        at processTicksAndRejections (node:internal/process/task_queues:93:5)
  Message:
    Expected 1 to be undefined.
  Stack:
    Error: Expected 1 to be undefined.
        at <Jasmine>
        at UserContext.<anonymous> (/Users/coreybaker/Documents/github/parse-server/spec/ParseQuery.spec.js:3317:43)
        at <Jasmine>
        at processTicksAndRejections (node:internal/process/task_queues:93:5)

2) Parse.Query testing exclude keys
  Message:
    Expected 'baz' to be undefined.
  Stack:
    Error: Expected 'baz' to be undefined.
        at <Jasmine>
        at UserContext.<anonymous> (/Users/coreybaker/Documents/github/parse-server/spec/ParseQuery.spec.js:3437:43)
        at <Jasmine>
        at processTicksAndRejections (node:internal/process/task_queues:93:5)
  Message:
    Expected 'world' to be undefined.
  Stack:
    Error: Expected 'world' to be undefined.
        at <Jasmine>
        at UserContext.<anonymous> (/Users/coreybaker/Documents/github/parse-server/spec/ParseQuery.spec.js:3438:45)
        at <Jasmine>
        at processTicksAndRejections (node:internal/process/task_queues:93:5)
```